Calculator: Add a Rounding menu

LucasChollet opened the pull request:
Last part of #11623 !

This patch adds a few ways to interact with those brand-new infinite precision numbers by adding some rounding abilities.

Finally, fixes #14062

LucasChollet commented on Userland/Applications/Calculator/main.cpp:
```diff
@@ -70,6 +69,31 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         widget->set_entry(Crypto::BigFraction { Crypto::SignedBigInteger(16180339887), power });
     }));
 
+    auto& round_menu = window->add_menu("&Round");
```
Sure, definitely better with a loop!
I kept 0-2-4 though

LucasChollet commented on Userland/Applications/Calculator/Keypad.h:
```diff
@@ -46,6 +48,8 @@ class Keypad final {
 
     mutable Crypto::BigFraction m_internal_value {};
 
+    Crypto::UnsignedBigInteger m_displayed_frac_length { 0 };
```
Honestly, I can't tell why I made this choice. Changed it to `unsigned`

trflynn89 commented:
So I tried this out locally, I think the UI needs some work. The dialog to set a custom rounding mode is quite squished vertically (looks like it's not even a full line height).

![Screenshot from 2022-09-24 08-41-27](https://user-images.githubusercontent.com/5600524/192098547-6af6b7ef-2d26-430e-9e80-87f172eaf52b.png)

It also appears in the middle of the desktop. It should be centered over the calculator window, and behave as a modal (i.e. I should not be able to interact with the calculator window while this is open).

I recommend looking at Minesweeper's "Custom game" dialog and implementing this dialog to match it (modal, a spin box to select values, cancel-able, etc.).

The casing on the confirmation button should match what's used everywhere else - that is, "OK" instead of "Ok".

trflynn89 commented:
That said, the more I use this, the more I think we're better off creating a settings dialog instead. The way we display the custom rounding back to the user seems a bit awkward, and the presets of 0, 2, 4 feel pretty arbitrary:
![Screenshot from 2022-09-24 08-55-06](https://user-images.githubusercontent.com/5600524/192099231-0440907f-42fe-4bab-b75d-cd28aefde4b4.png)

I think we should have a settings dialog instead of the radio buttons, and that dialog can have a spin box to set the rounding length. I think this would be much clearer, and will also make it easier to add more settings in the future.

LucasChollet commented:
My bad for the squished height, I thought that it was a bug in `LibGUI` but I was entirely my fault.

What I've done:
- [X] Adjusted the vertical size
- [X] The dialog is now modal
- [X] Centered the dialog
- [X] Added a cancel button
- [X] Ok :arrow_right:  OK
- [X] Replace the `TextEdit` widget with a `SpinBox`

The thing that led me to this design was quick user access. I wanted the user to be able to change the setting without going through too many clicks. Maybe we can keep the 0, 2, 4 and add a spinbox in the dialog for fine control? It's what I've pushed, tell me if you think that it feels wrong, and I will delete the radio buttons.

Having some Qt experience in the meantime definitely helps to understand `LibGUI` :sweat_smile:

trflynn89 commented:
IMO a settings app would still be better here, but this is significantly better than the last iteration, so we can move towards a settings app later :)

I think for the UI, the only thing that feels awkward still is that if I cancel the custom rounding length, the "Custom" option is still selected in the menu. Seems like the selection option shouldn't change when I cancel.

trflynn89 commented on Userland/Applications/Calculator/Keypad.cpp:
```diff
@@ -136,3 +134,8 @@ String Keypad::to_string() const
+
+void Keypad::set_rounding_length(unsigned rounding_threshold)
+{
+    m_displayed_frac_length = rounding_threshold;
```
We usually prefer using full names over abbreviations - `m_displayed_fraction_length`

trflynn89 commented on Userland/Applications/Calculator/main.cpp:
```diff
@@ -70,6 +69,22 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+    auto& round_menu = window->add_menu("&Round");
+    GUI::ActionGroup preview_actions;
+
+    static constexpr Array<StringView, 3> numbers_label { "zero"sv, "two"sv, "four"sv };
```
This is a minor, but I think it would feel more natural if these were "0", "2", and "4"? Using the full words for numbers seemed a tad off, IMO.

```c++
    static constexpr auto rounding_modes = Array { 0, 2, 4 };

    for (auto rounding_mode : rounding_modes) {
        auto round_action = GUI::Action::create_checkable(String::formatted("To &{} digits", rounding_mode), [&widget, rounding_mode](auto&) {
            widget->set_rounding_length(rounding_mode);
        });

        preview_actions.add_action(*round_action);
        round_menu.add_action(*round_action);
    }
```

trflynn89 commented:
There also seems to be a weird bug that unsets the rounding mode? To reproduce:

1. Open Calculator
2. Select the preset 2-digit rounding mode (Round > To Two digits)
3. In calculator, do some decimal math, e.g. `2.55555 + 1`
4. The custom rounding mode is set to 5 (the number of decimals in `2.55555`) instead of 2

LucasChollet commented:
> IMO a settings app would still be better here, but this is significantly better than the last iteration, so we can move towards a settings app later :)

Deal, you might have yakbait future me when I will think of it in one month :sweat_smile: 

> I think for the UI, the only thing that feels awkward still is that if I cancel the custom rounding length, the "Custom" option is still selected in the menu. Seems like the selection option shouldn't change when I cancel.

Should be fixed now!

> There also seems to be a weird bug that unsets the rounding mode? To reproduce:
> 
>     1. Open Calculator
> 
>     2. Select the preset 2-digit rounding mode (Round > To Two digits)
> 
>     3. In calculator, do some decimal math, e.g. `2.55555 + 1`
> 
>     4. The custom rounding mode is set to 5 (the number of decimals in `2.55555`) instead of 2

This is half-wanted behavior, I intentionally implemented a switch to custom mode when entering larger numbers but forgot to report it to the GUI. I tried to explain it in the commit message, is it unclear ?

There is now a callback to update the GUI, is it good for you?

trflynn89 commented:
> This is half-wanted behavior, I intentionally implemented a switch to custom mode when entering larger numbers but forgot to report it to the GUI. I tried to explain it in the commit message, is it unclear ?

Yeah this is very surprising behavior IMO. If I set 2 digits, I wouldn't expect that to be erased when I enter "1.2345" - I would expect the next computation to round to 2 digits.
